lib/config/providers/chat_search_provider.dart

```diff
@@ -0,0 +1,146 @@
+import 'dart:async';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:whitenoise/config/states/chat_search_state.dart';
+import 'package:whitenoise/domain/models/message_model.dart';
+
+final chatSearchProvider =
+    StateNotifierProvider.family<ChatSearchNotifier, ChatSearchState, String>(
+      (ref, groupId) => ChatSearchNotifier(groupId),
+    );
+
+class ChatSearchNotifier extends StateNotifier<ChatSearchState> {
+  final String groupId;
+  Timer? _debounceTimer;
+
+  ChatSearchNotifier(this.groupId) : super(const ChatSearchState());
+
+  void activateSearch() {
+    state = state.copyWith(isSearchActive: true);
+  }
+
+  void deactivateSearch() {
+    _debounceTimer?.cancel();
+    state = const ChatSearchState();
+  }
+
+  void updateQuery(String query) {
+    _debounceTimer?.cancel();
+
+    if (query.isEmpty) {
+      state = state.copyWith(
+        query: query,
+        matches: [],
+        currentMatchIndex: 0,
+        isLoading: false,
+      );
+      return;
+    }
+
+    state = state.copyWith(
+      query: query,
+      isLoading: true,
+    );
+
+    _debounceTimer = Timer(const Duration(milliseconds: 300), () {
+      _performSearch(query);
+    });
+  }
+
+  void _performSearch(String query) {
+    if (query.isEmpty) {
+      state = state.copyWith(
+        matches: [],
+        currentMatchIndex: 0,
+        isLoading: false,
+      );
+      return;
+    }
+
+    // This will be called from the chat screen with the actual messages
+    state = state.copyWith(isLoading: false);
+  }
+
+  void performSearchWithMessages(String query, List<MessageModel> messages) {
+    if (query.isEmpty) {
+      state = state.copyWith(
+        matches: [],
+        currentMatchIndex: 0,
+        isLoading: false,
+      );
+      return;
+    }
+
+    final allMatches = <SearchMatch>[];
+    final queryLower = query.toLowerCase();
+
+    // Process messages in normal order (newest first, as they appear in the list)
+    // But we want the topmost (oldest visible) match to be #1
+    for (int i = 0; i < messages.length; i++) {
+      final message = messages[i];
+      final content = message.content?.toLowerCase() ?? '';
+
+      if (content.contains(queryLower)) {
+        int startIndex = 0;
+
+        while (true) {
+          final index = content.indexOf(queryLower, startIndex);
+          if (index == -1) break;
+
+          // Create a separate SearchMatch for each individual word occurrence
+          allMatches.add(
+            SearchMatch(
+              messageId: message.id,
+              messageIndex: i,
+              messageContent: message.content ?? '',
+              textMatches: [
+                TextMatch(
+                  start: index,
+                  end: index + queryLower.length,
+                  matchedText: message.content!.substring(index, index + queryLower.length),
+                ),
+              ],
+            ),
+          );
+
+          startIndex = index + 1;
```

Comment:
Also a doubt here... shoudn't it start from `index + queryLower.length` ? For example, If Im searching "la" in a string "lalala", starts in 0, finds a match in 0, then the query has length 2 so it could start seaching from index 2 instead of 1 (And then continue searching in "lala" instead of "alala". Or am I missunderstanding how this works?